flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/StateMapSnapshot.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.state.heap;
+
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.core.memory.DataOutputView;
+import org.apache.flink.runtime.state.StateSnapshotTransformer;
+import org.apache.flink.util.Preconditions;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+
+/**
+ * Base class for snapshots of a {@link StateMap}.
+ *
+ * @param <K> type of key
+ * @param <N> type of namespace
+ * @param <S> type of state
+ */
+public abstract class StateMapSnapshot<K, N, S, T extends StateMap<K, N, S>> {
+
+	/**
+	 * The {@link StateMap} from which this snapshot was created.
+	 */
+	protected final T owningStateMap;
```

Comment:
This is only used by the copy on write variant and therefore could probably be removed from a common abstract class. The abstract class might then just be an interface.

Reply:
Actually, `owningStateMap ` is also used in `NestedStateMapSnapshot#writeState`.

Reply:
Right, sorry I overlooked that one :)